### PR TITLE
KAFKA-4302: Simplify KTableSource

### PR DESCRIPTION
**Pull request by mjsax**
KTableSource is always materialized since IQ:
- removed flag KTableSource#materialized
- removed MaterializedKTableSourceProcessor

### CODE REVIEW

**Comment by mjsax** [comment]
@guozhangwang @enothereska

**Comment by mjsax** [comment]
Build failed in know unstable test. Will trigger new build.

**Comment by enothereska** [comment]
LGTM

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java [hunk 1]
```diff
@@ -151,7 +153,14 @@ public KStreamBuilder() {
         addProcessor(name, processorSupplier, source);
 
         final KTableImpl kTable = new KTableImpl<>(this, name, processorSupplier, Collections.singleton(source), keySerde, valSerde, storeName);
-        kTable.materialize((KTableSource) processorSupplier);
+        StateStoreSupplier storeSupplier = new RocksDBKeyValueStoreSupplier<>(storeName,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:156]: We should also update the JavaDoc for this parameter, currently it is `the state store name used if this KTable is materialized, can be null if materialization not expected`.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:156]: Actually thinking about this a bit more, why do we always want to materialize the KTable source? If users do not provide a state store name when creating a KTable, it means they are not interested in querying it, so in this case we should probably not materialize it, no?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:156]: @enothereska Can you comment on this one? If we do not materialize always, this patch does not make sense all together.

Comment (enothereska) [reply streams/src/main/java/org/apache/kafka/streams/kstream/KStreamBuilder.java:156]: We decided to have all sources materialized when we had the Interactive Query design for 0.10.1. Before changing that, I'd like to get more feedback from the community first, so we don't go back and forth.

So I'd treat the Javadoc as incorrect and would be great if we can fix the Javadoc as part of this PR. Thanks @mjsax

**Comment by mjsax** [comment]
@guozhangwang @enothereska Updated.

**Comment by enothereska** [comment]
LGTM

**Comment by guozhangwang** [comment]
There is a potential issue that table-table-join resulted table does not have the `storeName`, and hence the following topology will cause throw exceptions:

```
stream.join(table1.join(table2))
```

on

```
topology.connectProcessorAndStateStores(name, other.getStoreName());
```

in `KStreamImpl`. @mjsax could you file another JIRA to keep track of it?

As for the solution, I think we may still need to materialize the KTable on-the-fly if the above scenario happens and `other.getStoreName()` returns null indicating that this KTable is not materialized yet.

**Comment by guozhangwang** [comment]
Merged to trunk.

**Pull request closed without merging**